src/psyclone/psyir/symbols/symboltable.py

```diff
@@ -180,40 +221,124 @@ def _normalize(key):
+        :rtype: :py:class:`psyclone.psyir.symbols.Symbol`
+        '''
+        try:
+            symbol = self.lookup_with_tag(tag)
+            # Check that the symbol found matches the requested description
+            if 'symbol_type' in new_symbol_args:
+                symbol_type = new_symbol_args['symbol_type']
+                if not isinstance(symbol, new_symbol_args['symbol_type']):
+                    raise SymbolError(
+                        "Expected symbol with tag '{0}' to be of type '{1}' "
+                        "but found type '{2}'.".format(
+                        tag, symbol_type.__name__, type(symbol).__name__))
+            # TODO: What happens if the symbol is found and some unmatching
+            # arguments were given? If it needs to fail, we can implement the
+            # previous check polymorphically inside each symbol:
+            # symbol.match(arguments)
+            # which can also be useful for symbol comparisons in general.
+            return symbol
+        except KeyError:
+            if not root_name:
+                root_name = tag
+            return self.new_symbol(root_name, tag, **new_symbol_args)
+
+    def next_available_name(self, root_name=None, shadowing=False):
+        '''Return a name that is not in the symbol table and therefore can
+        be used to declare a new symbol.
+        If the `root_name` argument is not supplied or if it is
         an empty string then the name is generated internally,
         otherwise the `root_name` is used. If required, an additional
         integer is appended to avoid clashes.
+        If the shadowing argument is True (is False by default), the names
+        in parent symbol tables will not be considered.
 
         :param root_name: optional name to use when creating a new \
             symbol name. This will be appended with an integer if the name \
             clashes with an existing symbol name.
         :type root_name: str or NoneType
-        :param bool check_ancestors: optional logical flag indicating \
-            whether the name should be unique in this symbol table \
-            (False) or in this and all ancestor symbol tables \
-            (True). Defaults to True.
+        :param bool shadowing: optional logical flag indicating whether the \
+            name can be overlapping with a symbol in any of the ancestors \
+            symbol tables. Defaults to False.
 
         :returns: the new unique symbol name.
         :rtype: str
 
         :raises TypeError: if the root_name argument is not a string \
             or None.
+        :raises TypeError: if the shadowing argument is not a bool.
 
         '''
-        if check_ancestors:
-            symbols = self._all_symbols
-        else:
+        if not isinstance(shadowing, bool):
+            raise TypeError("Argument shadowing should be of type bool"
+                " but found '{0}'.".format(type(shadowing).__name__))
+
+        if shadowing:
```

Comment:
Please add comment(s) here, e.g. "Symbol shadowing is not permitted so get symbols from all ancestor symbol tables"

Reply:
Done